src/sync/deduplicator.ts

```diff
@@ -0,0 +1,200 @@
+import {Redis} from "ioredis";
+
+/**
+ *
+ * The generic idea of the Deduplicator(tm) is very simple: while executing the job, Deduplicator raises a flag (in redis)
+ * to inform other executors not to run the same job ("same"-ness is defined by a key). At the end of the execution it
+ * removes the previously raised flag.
+ *
+ * As all "very simple" ideas, it has several challenges, namely:
+ *
+ * THE FIRST CHALLENGE is: what if the executor dies (e.g. the EC2 node is rotated) during the execution? This might
+ * have a flag raised but not being deleted.
+ *
+ * The implemented solution: save the timestamp of the flag and keep updating it in the executor while it is doing its
+ * job. If it dies, the flag will be invalidated in some reasonable time.
+ *
+ * THE SECOND CHALLENGE is: let's say an executor observes the flag, what should it do? Reschedule the job or drop it?
+ *
+ * Always rescheduling is bad: if we have the issue with duplicates, this will keep queue increasing, not good for Redis'
+ * CPU.
+ *
+ * Always dropping is dangerous: what if the flag belongs to a "dead" executor, however it wasn't invalidated yet?
+ *
+ * The implemented solution: once the executor observes the flag of a concurrent executor, it will watch for it for
+ * some period of time and if it is moving (timestamp is being updated, which means the executor is doing "something"),
+ * we are safe to discard the job as a detected duplicate; however if it is not moving it is dangerous to discard it
+ * and should be rescheduled instead.
```

Comment:
We have another option:
* Always discard if the flag is on
* Periodically scan the table for unprocessed syncs and then schedule a message if the flag is not there for that sync

Reply:
> Periodically scan the table for unprocessed syncs and then schedule a message if the flag is not there for that sync

TBH I don't like this suggestion. Here all the rescheduling logic is encapsulated in a single place and we don't need the database at all. In this suggestion we will have some periodic job.

And then there's a question which job should be considered as "staled", what's the criteria? A timeout? What is this is a long-running job?